Add "best practices" guidelines for upgrading Cumulus (#1633)
nasa/cumulus: docs/deployment/upgrade.md

After the initial deployment, any future updates to the Cumulus deployment from configuration files, Terraform files (`*.tf`), or modules from a new version of Cumulus can be deployed and will update the appropriate portions of the stack as needed.

## Cumulus Versioning

> Cumulus uses a global versioning approach, meaning version numbers are consistent across all packages and tasks, and semantic versioning to track major, minor, and patch version (e.g., 1.0.0). We use Lerna to manage versioning.

Reviewer: I realize this isn't a new content update, that said: 

` meaning version numbers are consistent across all packages and tasks,`

I'm a little concerned about this - this is only true for our @cumulus/package npm package.    Users might infer *other* packages (e.g. CMA client libs, cumulus-ecs-task, etc ) are also synced which is most definitely not true.  

I'm wondering if we can be explicit here in what we mean when we say packages and tasks somehow...  I'd suggest a link to the NPM repository, but I'm advocating we take down *all* of our tasks from npm publication and probably a good number of our packages as well. 

More to the point, unless they're writing lambdas users shouldn't ever need to care about this -- they're *only* pulling in the TF module at this point. 

I think the requirements in 1967 (All modules should be on the same version - i.e. data-persistence and cumulus) speak to this a little.     I'm torn though, as we're currently publishing those modules, it's just users shouldn't be using them at all anymore outside of node task/core development, for the most part.

Reviewer 2: I agree that we need to update this to make more sense in the Terraform world. Since we're talking about deployment updates, we probably don't need to mention the packages too much.

Reviewer 2: I also think we can take out the sentence `We use Lerna to manage versioning.`

Author: I'll take out the bit about Lerna.

Regarding wording around version number consistency, how about something like this:

> meaning version numbers are consistent across all Terraform modules

Reviewer 2: 👍

To view the released module artifacts for each Cumulus core version, see the [Cumulus Releases] page.

> **IMPORTANT:** Cumulus releases may introduce breaking changes between minor versions (e.g., 1.13.x -> 1.14.0), so it is critical that you consult the release notes for migration steps.

Reviewer: I think we should be more explicit that our minor versions *are* breaking change versions.   We shouldn't be upping the minor version unless there is a breaking change.     Something along the lines of, but not precisely:  

```suggestion
**IMPORTANT:** By convention, Cumulus minor version releases introduce breaking changes (e.g., 1.13.x -> 1.14.0), so it is critical that you consult the release notes for migration steps.
```

It might be a good idea to also explicitly note that our x.y.z format doesn't follow SemVer, but that's more of a personal preference than a generally applicable PR comment.

Reviewer 2: I agree with Jonathan, but I think that should go up to the first paragraph that says we use semantic versioning.

## Migrating to a New Version

> When breaking changes have been introduced, the Cumulus Core team will publish instructions on migrating from one version to another.  Detailed release notes with migration instructions (if any) for each release can be found on the [Cumulus Releases] page.

Reviewer: Should we mention the 'breaking changes' section in the CHANGELOG here?

Author: Ah, yes. Good call.

Author: How would you feel about adding a mention regarding deprecations as well? Something to the effect that this might be a good time to replace calls to deprecated functions?

Reviewer 2: I like the idea to mention deprecations, but we need to make the distinction that's only needed if you're using functions from the packages. If you're just using TF deployments that does not affect you. Maybe a separate section about updating your code using Cumulus packages from npm?

Reviewer: Definitely something noting that: 

- Your workflow lambdas that utilize published Core npm modules should always match your deployed Cumulus version to ensure compatibility
- Check the CHANGELOG for deprecation/breaking change notices

**IMPORTANT** 

1. **Use consistent Cumulus versions:** All modules must be updated to the same Cumulus version number (see below)
2. **Follow all intervening steps:** When skipping over versions, you **must perform all intervening migration steps**.  For example, if going from version 1.1.0 to 1.3.0, upgrade from 1.1.0 to 1.2.0 and then to 1.3.0.  This is critical because each release that contains migration steps provide instructions _only_ for migrating from the _immediately_ previous release, but you must follow _all_ migration steps between your currently installed release and _every release_ through the release that you wish to migrate to.

> 3. **Migrate lower environments first:** Migrate your "lowest" environment first and test it to ensure correctness before performing migration steps in each successively higher environment.  For example, in the EDCloud progress from a sandbox environment to UAT to SIT to PROD.

Reviewer: > For example, in the EDCloud progress from a sandbox environment to UAT to SIT to PROD.

This might be confusing to users who don't grok what EDCloud is - can we maybe use a different descriptor?

Reviewer: Perhaps it doesn't matter as you already made the point in the previous sentence.

Reviewer: Also in the ticket: 

>  guidance on what is a good smoke test? Mention examples of use of EDCloud UAT vs. SIT vs. PROD environments.

I think based on that we need to add content on how to determine things were upgraded properly, or decide explicitly not to add that.

Reviewer 2: I think we can be generic here, like 'before updating your production environment, be sure to update lower testing environments, for example update Sandbox, then SIT, then UAT, the Prod'. Also we should encourage them to do some kind of smoke test. What's a good smoke test? Run through an ingest workflow?

Author: I'll remove mention of EDCloud.

Any suggestions on basic smoke tests to insert here?

Reviewer 2: I think run an ingest workflow is good

Reviewer: The actuality of a 'smoke test' for a componentized system is .... weird.

The problem with an ingest workflow is a user who has a default deployment....doesn't have one.    But ignoring that, evaluating components: 

- API works
- Workflow run/things appear in api? (yikes in production though) 
- distribution works?
- ....?

I'm not sure I'm advocating for this, just noting it was in the ticket.   We might be fine leaving that up to individual users given the widely divergent configuration options (that get bigger the better we do in moving toward components)

> 4. **Migrate during appropriate times:** Choose a time to migrate when support is more likely to be available in case you encounter problems, such as when you are most likely to quickly find help in the #cumulus-internal Slack channel.  Prefer earlier in the week over later in the week (particularly avoiding Fridays, if possible).

Reviewer: ```suggestion
4. **Migrate during appropriate times:** Choose a time to migrate when support is more likely to be available in case you encounter problems, such as when you are most likely to be able to obtain support.  Prefer earlier in the week over later in the week (particularly avoiding Fridays, if possible).
```

Again, generic audience concern.    Perhaps more importantly that location might change in the future, and I'm certain we won't remember to update this. 

 *Probably* not a big deal, but if we start seeing issues from external entities asking where our slack channel is and how they get support.... :D

Reviewer 2: Yep, agree. We should leave the Slack channel out of this docs.

Author: Makes sense.

## Updating Cumulus Version

To update your Cumulus version:

1. Find the desired release on the [Cumulus Releases] page
2. Update the `source` in your Terraform deployment files **for each of [your Cumulus modules](./components.md#available-cumulus-components)** by replacing `vx.x.x` with the desired version of Cumulus.  For example, here's the
entry from the `data-persistence` module:

    `source = "https://github.com/nasa/cumulus/releases/download/vx.x.x/terraform.zip//tf-modules/data-persistence"`

3. Run `terraform init` to get the latest copies of your updated modules

> ## Update Data Persistence Resources

Reviewer: Capitalizing these headers is inconsistent with how we've been updating our docs recently (we've updated PRs the other way ... often)   We should have a conversation about the right standard.     No strong feelings from me other than ensuring consistency moving forward.

Author: No worries. I like to be consistent, so I'll revert this change. (As an aside, I do prefer title-cased titles, but consistency trumps preferences.)

**Reminder:** Remember to [initialize Terraform](./README.md#initialize-terraform) if necessary.

From the directory of your `data-persistence` deployment module (e.g., `data-persistence-tf`):

```bash
$ AWS_REGION=<region> \ # e.g. us-east-1
    AWS_PROFILE=<profile> \
    terraform apply
```

## Update Cumulus Resources

**Reminder:** Remember [to initialize Terraform](./README.md#initialize-terraform) if necessary.

From the directory of your `cumulus` deployment module (e.g., `cumulus-tf`):

```bash
$ AWS_REGION=<region> \ # e.g. us-east-1
    AWS_PROFILE=<profile> \
    terraform apply
```

> 

Reviewer 2: Maybe add sections for updating the dashboard if applicable to take advantage of latest API functionality and a section for verifying your Cumulus deployment.

Author: I don't know what that would look like, so I'm happy to take suggestions.

Reviewer 2: Something along the lines of: After deployment and migrations, run tests to verify your workflows are working before promoting the changes to higher environments. A good test to run would be your discovery and ingest workflows to verify ingest has not been impacted by the deployment.

Author: Ah, right. I added such details. I thought this comment might be referring to something else. Please see my other changes, which address this.

Reviewer 2: Yep can you just add it as a step here please? You can refer back to the part about smoke tests.
